# Hola Juan!


Mi nombre es David Bautista, soy code reviewer de Tripleten y hoy tengo el gusto de revisar tu proyecto.

Cuando vea un error la primera vez, lo señalaré. Deberás encontrarlo y arreglarlo. La intención es que te prepares para un espacio real de trabajo. En un trabajo, el líder de tu equipo hará lo mismo. Si no puedes solucionar el error, te daré más información en la próxima ocasión.

Encontrarás mis comentarios más abajo - por favor, no los muevas, no los modifiques ni los borres.

¿Cómo lo voy a hacer? Voy a leer detenidamente cada una de las implementaciones que has llevado a cabo para cumplir con lo solicitado. Verás los comentarios de esta forma:

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si todo está perfecto.
</div>


<div class="alert alert-block alert-warning">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Si tu código está bien pero se puede mejorar o hay algún detalle que le hace falta.
</div>


<div class="alert alert-block alert-danger">
    
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
    
Si de pronto hace falta algo o existe algún problema con tu código o conclusiones.
</div>


Puedes responderme de esta forma: 

<div class="alert alert-block alert-info">
<b>Respuesta del estudiante</b> <a class="tocSkip"></a>
</div>

¡Empecemos!

# Déjame escuchar la música

# Contenido <a id='back'></a>

* [Introducción](#intro)
* [Etapa 1. Descripción de los datos](#data_review)
    * [Conclusiones](#data_review_conclusions)
* [Etapa 2. Preprocesamiento de datos](#data_preprocessing)
    * [2.1 Estilo del encabezado](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusiones](#data_preprocessing_conclusions)
* [Etapa 3. Prueba de hipótesis](#hypotheses)
    * [3.1 Hipótesis 1: actividad de los usuarios y las usuarias en las dos ciudades](#activity)
    * [3.2 Hipótesis 2: preferencias musicales los lunes y los viernes](#week)
    * [3.3 Hipótesis 3: preferencias de género en Springfield y Shelbyville](#genre)
* [Conclusiones](#end)

## Introducción <a id='intro'></a>
Como analista de datos, tu trabajo consiste en analizar datos para extraer información valiosa y tomar decisiones basadas en ellos. Esto implica diferentes etapas, como la descripción general de los datos, el preprocesamiento y la prueba de hipótesis.

Siempre que investigamos, necesitamos formular hipótesis que después podamos probar. A veces aceptamos estas hipótesis; y otras veces, las rechazamos. Para tomar las decisiones correctas, una empresa debe ser capaz de entender si está haciendo las suposiciones correctas.

En este proyecto, compararás las preferencias musicales de las ciudades de Springfield y Shelbyville. Estudiarás datos reales de transmisión de música online para probar las hipótesis a continuación y comparar el comportamiento de los usuarios y las usuarias de estas dos ciudades.

### Objetivo:
Prueba tres hipótesis:
1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.
2. Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre los viernes por la noche.
3. Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop, mientras que en Shelbyville hay más personas a las que les gusta el rap.

### Etapas
Los datos del comportamiento del usuario se almacenan en el archivo `/datasets/music_project_en.csv`. No hay ninguna información sobre la calidad de los datos, así que necesitarás examinarlos antes de probar las hipótesis.

Primero, evaluarás la calidad de los datos y verás si los problemas son significativos. Entonces, durante el preprocesamiento de datos, tomarás en cuenta los problemas más críticos.

Tu proyecto consistirá en tres etapas:
 1. Descripción de los datos
 2. Preprocesamiento de datos
 3. Prueba de hipótesis


### Desafío

En este proyecto, preparamos un pequeño reto para ti. Incluimos un nuevo tipo de estructura de datos: las marcas temporales. Las marcas temporales son muy comunes y merecen una atención adicional. Más adelante en el programa, aprenderás mucho sobre ellas. Sin embargo, por ahora las trataremos como simples strings. Necesitamos marcas temporales en este proyecto para poner a prueba una de nuestras hipótesis. No te preocupes, te ayudaremos con esto. Tu nivel de conocimientos actual será suficiente para abordarlo.

Por ejemplo, digamos que tenemos dos marcas temporales: `dt1 = "12:00:00"` y `dt2 = "06:00:00"`. Queremos comparar estas dos marcas temporales y ver cuál es posterior.

Podemos compararlas mediante los operadores de comparación estándar (`<`, `>`, `<=`, `>=`, `==`, `!=`). Ejecuta la siguiente celda de código para comparar dos marcas temporales:


In [1]:
# Comparar los objetos datetime

dt1 = "12:00:00"
dt2 = "06:00:00"

if dt1 < dt2:
    print("La marca temporal 2 es posterior")
else:
    print("La marca temporal 1 es posterior")

La marca temporal 1 es posterior


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Muy bien, recuerda que para todo los proyectos que realices es importante genera un contexto donde se comente que trata el caso y cuáles con los objetivos a cumplir, además es indispensable que se genere una tabla de contenido con el fin de mantener el orden en la entrega. Buen trabajo con el reto.
</div>

[Volver a Contenidos](#back)

## Etapa 1. Descripción de los datos <a id='data_review'></a>

Abre los datos en Y.Music y examínalos.

Necesitarás `pandas`, así que impórtalo.

In [2]:
import pandas as pd # importar pandas


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo importando ``pandas``.
</div>

Lee el archivo `music_project_en.csv` de la carpeta `/datasets/` y guárdalo en la variable `df`:

In [3]:
path = '/datasets/music_project_en.csv'
df = pd.read_csv(path)# Leer el archivo y almacenarlo en df

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo cargando los datos.
</div>

Muestra las 10 primeras filas de la tabla:

In [4]:
df.head(10)


userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5  Shelbyville  13:09:41     Friday  
6  Springfield  13:00:07  Wednesday  
7  Springfield  20:47:49  Wednesday  
8  Springfield  09:17:40     Friday  
9  Shelbyville  21:20:49  Wednesday

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con el ``head()``
</div>

Obtén la información general sobre la tabla con un comando:

In [5]:
# también puede ejecutarse individualmente los atributos
# .shape , .columns , .size , .index ,etc.

print(df.info())
print('--------------------------')  #se usa este print para delimitar espacios y permitir mejor lectura
print(df.columns)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0     userID  65079 non-null  object
 1   Track     63736 non-null  object
 2   artist    57512 non-null  object
 3   genre     63881 non-null  object
 4     City    65079 non-null  object
 5   time      65079 non-null  object
 6   Day       65079 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB
None
--------------------------
Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


In [6]:
duplicated_data = df.duplicated().sum()
na_data = df.isna().sum()

print(f'Tienes {duplicated_data} valores duplicados.')
print('Tienes los siguieres valores ausentes: \n', na_data)


Tienes 3826 valores duplicados.
Tienes los siguieres valores ausentes: 
   userID       0
Track       1343
artist      7567
genre       1198
  City         0
time           0
Day            0
dtype: int64


Estas son nuestras observaciones sobre la tabla. Contiene siete columnas. Todas almacenan el mismo tipo de datos: `object` (objeto).

Según la documentación:
- `' userID'` — identificador del usuario o la usuaria;
- `'Track'` — título de la canción;
- `'artist'` — nombre del artista;
- `'genre'` — género musical;
- `'City'` — ciudad del usuario o la usuaria;
- `'time'` — hora exacta en la que se reprodujo la canción;
- `'Day'` — día de la semana.

Podemos ver tres problemas con el estilo en los encabezados de la tabla:
1. Algunos encabezados están en mayúsculas; otros, en minúsculas.
2. Hay espacios en algunos encabezados.
3. Palabras unidas en lugar de usar snake_case. - Respuesta



### Tus observaciones <a id='data_review_conclusions'></a>

`Escribe tus observaciones aquí:

`1.   ¿Qué tipo de datos tenemos a nuestra disposición en las filas? ¿Y cómo podemos entender lo que almacenan las columnas?`
- Tenemos clasificados 7 columnas de datos tipo object y podemos entenderlos mediante el comando info() y/o leyendo la documentación. También puede ayudarnos correr un head() en el conjunto de datos para vericar que información podemos visualizar 

`2.   ¿Hay suficientes datos para proporcionar respuestas a nuestras tres hipótesis o necesitamos más información?`
- Sí, contamos con las columnas necesarias, haciendo enfasis en 'genre', 'day', 'city' y 'time'

`3.   ¿Notaste algún problema en los datos, como valores ausentes, duplicados o tipos de datos incorrectos?`
- Si, hay valores ausentes en las columnas "Track", 'artist' ,"genre" y tenemos 3826 valores duplicados, valores incorectos en los headers

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Genial Juan. Muy buen trabajo con la sección.
</div>

[Volver a Contenidos](#back)

## Etapa 2. Preprocesamiento de datos <a id='data_preprocessing'></a>

El objetivo aquí es preparar los datos para que sean analizados.
El primer paso es resolver cualquier problema con los encabezados. Luego podemos avanzar a los valores ausentes y duplicados. Empecemos.

Corrige el formato en los encabezados de la tabla.


### Estilo del encabezado <a id='header_style'></a>
Muestra los encabezados de la tabla:

In [7]:
print(df.columns)


Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='object')


Cambia los encabezados de la tabla de acuerdo con las reglas del buen estilo:
* todos los caracteres deben ser minúsculas;
* elimina los espacios;
* si el nombre tiene varias palabras, utiliza snake_case.

* Pon todos los caracteres en minúsculas y muestra el encabezado de la tabla de nuevo. 

* Ahora elimina los espacios al principio y al final de los encabezados y muéstralos.

* Aplica snake_case al encabezado userID y muestra el encabezado de la tabla.

* Comprueba el resultado. Muestra los encabezados una vez más.

In [8]:
data_columns = df.columns 
new_columns = [] #se configura a nivel lista para acceder desde df.columns, de lo contrario se tendría que acceder como diccionario

for column in data_columns:
    lower_data = column.lower()
    strip_data = lower_data.strip()
    replace_data = strip_data.replace(' ','_')
    if replace_data == 'userid': #se ofrece opción adicional en el siguiente comentario
        replace_data = 'user_id'
    new_columns.append(replace_data)

df.columns = new_columns

# df.rename(columns={'userid':'user_id'}, inplace=True) # Otra forma de cambiar userid a user_id

print(df.columns)
    


Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='object')


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo Juan, muy buena forma de solucionar el problema del nombre en las columnas.
</div>

[Volver a Contenidos](#back)

### MissValores ausentes <a id='missing_values'></a>
Primero, encuentra el número de valores ausentes en la tabla. Para ello, utiliza dos métodos pandas:

In [9]:
null_values = df.isna().sum() # Para calcular el número de valores ausentes por columnas
print(null_values)

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


No todos los valores ausentes afectan a la investigación. Por ejemplo, los valores ausentes en `track` y `artist` no son cruciales. Simplemente puedes reemplazarlos con valores predeterminados como el string `'unknown'` (desconocido).

Pero los valores ausentes en `'genre'` pueden afectar la comparación entre las preferencias musicales de Springfield y Shelbyville. En la vida real, sería útil saber las razones por las cuales hay datos ausentes e intentar recuperarlos. Pero no tenemos esa oportunidad en este proyecto. Así que tendrás que:
* rellenar estos valores ausentes con un valor predeterminado;
* evaluar cuánto podrían afectar los valores ausentes a tus cómputos;

Reemplazar los valores ausentes en `'track'`, `'artist'` y `'genre'` con el string `'unknown'`. Para hacer esto, crea la lista `columns_to_replace`, recórrela con un bucle `for` y reemplaza los valores ausentes en cada columna:

In [10]:
# Bucle en los encabezados reemplazando los valores ausentes con 'unknown'

columns_to_replace = ['track','artist', 'genre']

for column in columns_to_replace:
    df[column].fillna('unknown', inplace=True)

# new_null_values = df.isna().sum()
# print(new_null_values)
# print(df.head())
    

Asegúrate de que la tabla no contiene más valores ausentes. Cuenta de nuevo los valores ausentes.

In [11]:
new_null_values = df.isna().sum()
print(new_null_values)

user_id    0
track      0
artist     0
genre      0
city       0
time       0
day        0
dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo lidiando con los valores nulos.</div>

[Volver a Contenidos](#back)

### Duplicados <a id='duplicates'></a>
Encuentra el número de duplicados explícitos en la tabla usando un comando:

In [12]:
data_duplicated = df.duplicated().sum()

print(f'La base cuenta con {data_duplicated} valores duplicados')


La base cuenta con 3826 valores duplicados


Llama al método `pandas` para deshacerte de los duplicados explícitos:

In [13]:
df.drop_duplicates(inplace=True)

Cuenta los duplicados explícitos una vez más para asegurarte de haberlos eliminado todos:

In [14]:
result = df.duplicated().sum()

#Corrección primera revisión
df = df.drop_duplicates().reset_index(drop=True) #Reseteando el index, sin agregarla como columna
print(result)

0


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~Has realizado un buen trabajo. No obstante, es importante que consideres que al eliminar los duplicados se puede generar un problema con la indexación de la tabla. Para resolverlo, puedes seguir este ejemplo:~~
    
```python
df = df.drop_duplicates().reset_index(drop=True)    
```    
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo.</div>

Ahora queremos deshacernos de los duplicados implícitos en la columna `genre`. Por ejemplo, el nombre de un género se puede escribir de varias formas. Dichos errores también pueden afectar al resultado.

Para hacerlo, primero mostremos una lista de nombres de género únicos, ordenados en orden alfabético. Para hacerlo:
* recupera la columna deseada del dataFrame;
* llama al método que te devolverá todos los valores de columna únicos;
* aplica un método de ordenamiento a tu resultado.


In [15]:
genre_uniques = df['genre']

genre_uniques = genre_uniques.sort_values().unique()

for genre in genre_uniques: #aplico esta formula para visualmente poder analizarlo mejor en lugar del formato lista
    print(genre)


acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'heavy
hardcore
hardstyle
hardtechno
hip
hip-hop
hiphop
historisch
holiday
hop
horror
house
idm
i

Busca en la lista para encontrar duplicados implícitos del género `hiphop`. Estos pueden ser nombres escritos incorrectamente o nombres alternativos para el mismo género.

Verás los siguientes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para deshacerte de ellos, declara la función `replace_wrong_genres()` con dos parámetros:
* `wrong_genres=` — la lista de duplicados;
* `correct_genre=` — el string con el valor correcto.

La función debería corregir los nombres en la columna `'genre'` de la tabla `df`, es decir, remplaza cada valor de la lista `wrong_genres` con el valor en `correct_genre`. Utiliza un bucle `'for'` para iterar sobre la lista de géneros incorrectos y reemplazarlos con el género correcto en la lista principal.

In [16]:
# Función para reemplazar duplicados implícitos
# .replace(old_element,new_element) - remplazar

wrong_genres= ["hip",'hop','hip-hop']
correct_genre= 'hiphop'

def replace_wrong_genres(wrong_genres,correct_genre):
    for word in wrong_genres:
        df['genre'] = df['genre'].replace(word,correct_genre)


Llama a `replace_wrong_genres()` y pásale argumentos para que retire los duplicados implícitos (`hip`, `hop` y `hip-hop`) y los reemplace por `hiphop`:

In [17]:
replace_wrong_genres(wrong_genres,correct_genre)



Asegúrate de que los nombres duplicados han sido eliminados. Muestra la lista de valores únicos de la columna `'genre'` una vez más:

In [18]:
clean_implicit = df['genre'].sort_values().unique()
# df['genre'].sort_values().nunique()

for genre in clean_implicit: #aplico esta formula para visualmente poder analizarlo mejor en lugar del formato lista
    print(genre)


acid
acoustic
action
adult
africa
afrikaans
alternative
ambient
americana
animated
anime
arabesk
arabic
arena
argentinetango
art
audiobook
avantgarde
axé
baile
balkan
beats
bigroom
black
bluegrass
blues
bollywood
bossa
brazilian
breakbeat
breaks
broadway
cantautori
cantopop
canzone
caribbean
caucasian
celtic
chamber
children
chill
chinese
choral
christian
christmas
classical
classicmetal
club
colombian
comedy
conjazz
contemporary
country
cuban
dance
dancehall
dancepop
dark
death
deep
deutschrock
deutschspr
dirty
disco
dnb
documentary
downbeat
downtempo
drum
dub
dubstep
eastern
easy
electronic
electropop
emo
entehno
epicmetal
estrada
ethnic
eurofolk
european
experimental
extrememetal
fado
film
fitness
flamenco
folk
folklore
folkmetal
folkrock
folktronica
forró
frankreich
französisch
french
funk
future
gangsta
garage
german
ghazal
gitarre
glitch
gospel
gothic
grime
grunge
gypsy
handsup
hard'n'heavy
hardcore
hardstyle
hardtechno
hiphop
historisch
holiday
horror
house
idm
independent
india

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo solucionando el problema de los valores duplicados no tan obvios.
</div>

[Volver a Contenidos](#back)

### Tus observaciones <a id='data_preprocessing_conclusions'></a>

Descubrí que había 4 elementos relacionados al genero solicitado 'hip, hop, hip-hop, hiphop', se corrigio la base conservando un solo nombre. También encontre los siguientes elementos 'latin' y 'latino', sin embargo, al no tener completa veracidad de que es un genero duplicado decidí no procesarla, dandome un total de 266 generos.


[Volver a Contenidos](#back)

## Etapa 3. Prueba de hipótesis <a id='hypotheses'></a>

### Hipótesis 1: comparar el comportamiento del usuario o la usuaria en las dos ciudades <a id='activity'></a>

La primera hipótesis afirma que existen diferencias en la forma en que los usuarios y las usuarias de Springfield y Shelbyville consumen música. Para comprobar esto, usa los datos de tres días de la semana: lunes, miércoles y viernes.

* Agrupa a los usuarios y las usuarias por ciudad.
* Compara el número de canciones que cada grupo reprodujo el lunes, el miércoles y el viernes.


Realiza cada cálculo por separado.

Para evaluar la actividad de los usuarios y las usuarias en cada ciudad, agrupa los datos por ciudad y encuentra la cantidad de canciones reproducidas en cada grupo.



In [19]:
# Contando las canciones reproducidas en cada ciudad
#Los valores se repetirán ya que al ser todos valores object solo es posible usar count, 
# de lo contrario concatenaría la información de cada celda. Es posible limitar a una columna 
# utilizando df.groupby('city')['other_column'].count()

city = df.groupby('city')['track'].count() 


print(city)


city
Shelbyville    18512
Springfield    42741
Name: track, dtype: int64


Ahora agrupa los datos por día de la semana y encuentra el número de canciones reproducidas el lunes, miércoles y viernes.


In [20]:
# calculando las canciones reproducidas en cada uno de los tres días

city = df.groupby('day').count()['track']

shelby_city = df[df['city']=='Shelbyville']
shelby_city = shelby_city.groupby('day').count()['track']

springfield_city = df[df['city']=='Springfield']
springfield_city = springfield_city.groupby('day').count()['track']


print('\n Canciones por día a Total')
print(city)
print('\n Canciones por día en Shelbyville')
print(shelby_city)
print('\n Canciones por día en Springfield')
print(springfield_city)



 Canciones por día a Total
day
Friday       21840
Monday       21354
Wednesday    18059
Name: track, dtype: int64

 Canciones por día en Shelbyville
day
Friday       5895
Monday       5614
Wednesday    7003
Name: track, dtype: int64

 Canciones por día en Springfield
day
Friday       15945
Monday       15740
Wednesday    11056
Name: track, dtype: int64


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo con estas primeras agrupaciones.
</div>

He usado los filtros para poder accesar a cada una de las ciudades y extraér los días con la funcion groupby, de esta manera puedo mostrar tres tablas, a nivel general y por cada ciudad.

Ya sabes cómo contar entradas agrupándolas por ciudad o día. Ahora necesitas escribir una función que pueda contar entradas según ambos criterios simultáneamente.

Crea la `number_tracks()` para calcular el número de canciones reproducidas en un determinado día y ciudad. La función debe aceptar dos parámetros:

- `day`: un día de la semana para filtrar. Por ejemplo, `'Monday'`.
- `city`: una ciudad para filtrar. Por ejemplo, `'Springfield'`.

Dentro de la función, aplicarás un filtrado consecutivo con indexación lógica.

Primero filtra los datos por día y luego filtra la tabla resultante por ciudad.

Después de filtrar los datos por dos criterios, cuenta el número de valores de la columna 'user_id' en la tabla resultante. Este recuento representa el número de entradas que estás buscando. Guarda el resultado en una nueva variable y devuélvelo desde la función.

In [21]:

day = 'Monday'
city = 'Shelbyville'

# def number_tracks(day,city):
#     track_list = df[df['day']== day] #Se asigna el filtro del día especifico
#     track_list = track_list[track_list['city']== city] #Se almacena el filtro de ciudad sobre el previo filtro de día
#     track_list_count = track_list['user_id'].count()
#     return track_list_count

#Corrección primera revisión
def number_tracks(day,city): #nueva version de la funcion donde se anidan las condciones de 
    track_list = df[(df['day']==day) & (df['city']==city)]
    track_list_count = track_list['user_id'].count()
    return track_list_count

print(number_tracks(day,city)) #Comprobando resultado acorde a las tablas anteriores.

5614


Llama a `number_tracks()` seis veces, cambiando los valores de los parámetros para que recuperes los datos de ambas ciudades para cada uno de los tres días.

In [22]:
# El número de canciones reproducidas en Springfield el lunes
day = 'Monday'
city = 'Springfield'
print(number_tracks(day,city))

15740


In [23]:
# El número de canciones reproducidas en Shelbyville el lunes
day = 'Monday'
city = 'Shelbyville'
print(number_tracks(day,city))

5614


In [24]:
# El número de canciones reproducidas en Springfield el miércoles
day = 'Wednesday'
city = 'Springfield'
print(number_tracks(day,city))

11056


In [25]:
# El número de canciones reproducidas en Shelbyville el miércoles
day = 'Wednesday'
city = 'Shelbyville'
print(number_tracks(day,city))

7003


In [26]:
# El número de canciones reproducidas en Springfield el viernes
day = 'Friday'
city = 'Springfield'
print(number_tracks(day,city))

15945


In [27]:
# El número de canciones reproducidas en Shelbyville el viernes
day = 'Friday'
city = 'Shelbyville'
print(number_tracks(day,city))

5895


Utiliza `pd.DataFrame` para crear una tabla, donde
* Los encabezados de la tabla son: `['city', 'monday', 'wednesday', 'friday']`
* Los datos son los resultados que conseguiste de `number_tracks()`

In [28]:
# Tabla con los resultados

headers = ['city', 'monday', 'wednesday', 'friday']
base = [['Shelbyville',5614,7003,5895],['Springfield',15740,11056,15945]]

new_dataframe = pd.DataFrame(data = base, columns=headers)

print(new_dataframe)

          city  monday  wednesday  friday
0  Shelbyville    5614       7003    5895
1  Springfield   15740      11056   15945


**Conclusiones**

Dentro del registro de personas activas que tenemos en nuestro conjunto de datos, tenemos un total de __18,512 personas en Shelbyville__ y __42,741 en Springfield__. Será necesario trabajar con valores porcentuales para poder comprobar nuestra teoría.



In [29]:
# Tabla con los resultados

headers = ['city', 'monday', 'wednesday', 'friday','total']
base = [['Shelbyville',30,38,32,100],['Springfield',37,26,37,100]]

new_dataframe = pd.DataFrame(data = base, columns=headers)

print(new_dataframe)

          city  monday  wednesday  friday  total
0  Shelbyville      30         38      32    100
1  Springfield      37         26      37    100


Los días de mayor consumo de canciones para Shelbyville son los día miercoles, mientras que para springfield los días de mayor consumo suelen efectuarse los viernes seguido de los días Lunes. Con esto podemos definir que los días de consumo son diferentes en estas ciudades y podemos aprobar la hipotesis.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~Buen trabajo con la funcion pero podrias tratar de anidar las condiciones de la tabla, te dare una idea.~~
    
```python
# Filtramos los registros de df donde day es monday y city es san francisco
df[(df['day']=='monday') & (df['city']=='san francisco')]
```    
</div>


<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Genial, muy buen trabajo Juan.</div>

[Volver a Contenidos](#back)

### Hipótesis 2: música al principio y al final de la semana <a id='week'></a>

Según la segunda hipótesis, el lunes por la mañana y el viernes por la noche, los ciudadanos de Springfield escuchan géneros diferentes a los que disfrutan los usuarios de Shelbyville.

Cree dos tablas con los nombres proporcionados en los dos bloques de código siguientes:
* Para Springfield — `spr_general`
* Para Shelbyville — `shel_general`

In [57]:
spr_general = df[df['city']=='Springfield']
# print(spr_general.head()) #comprobado

        user_id                          track                   artist  \
1      55204538    Delayed Because of Accident         Andreas Rönnberg   
4      E2DC1FAE                    Soul People               Space Echo   
6      4CB90AA5                           True             Roman Messer   
7      F03E1C1F               Feeling This Way          Polina Griffith   
8      8FA1D3BE                       L’estate              Julia Dalia   
...         ...                            ...                      ...   
61247  83A474E7  I Worship Only What You Bleed  The Black Dahlia Murder   
61248  729CBB09                        My Name                   McLean   
61250  C5E3A0D5                      Jalopiina                  unknown   
61251  321D0506                  Freight Train            Chas McDevitt   
61252  3A64EF84      Tell Me Sweet Little Lies             Monica Lopez   

              genre         city      time        day  
1              rock  Springfield  14:07:09 

In [31]:
shel_general = df[df['city']=='Shelbyville']
# print(shel_general.head()) #comprobado

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo creando las tablas.
</div>

Escribe la función `genre_weekday()` con cuatro parámetros:
* Una tabla para los datos (`df`)
* El día de la semana (`day`)
* La primera marca de tiempo, en formato (`time1`)
* La última marca de tiempo, en formato (`time2`)

La función debería devolver información de los 15 géneros más populares de un día determinado en un período entre dos marcas de fecha y hora.
Aplique la misma lógica de filtrado consecutivo, pero esta vez utilice cuatro filtros y luego cree una nueva columna con los respectivos recuentos de reproducción.
Ordene el resultado de mayor a menor y devuélvalo.

In [32]:

# def genre_weekday(df,day,time1,time2):
#     genre_df= df[df['day']==day]
#     genre_df= genre_df[genre_df['time']>time1]
#     genre_df= genre_df[genre_df['time']<time2]
#     genre_df_count = genre_df.groupby('genre')['track'].count() #size() #agregar columna, pendiente...
#     genre_df_sorted = genre_df_count.sort_values(ascending=False)
#     return genre_df_sorted[:15]

#Corrección primera revisión
def genre_weekday(df,day,time1,time2):
    genre_df= df[(df['day']==day) & (df['time']>time1) & (df['time']<time2)] #Se anidan todas las condiciones.
    genre_df_count = genre_df.groupby('genre')['track'].count()
    genre_df_sorted = genre_df_count.sort_values(ascending=False)
    return genre_df_sorted[:15]





    

Compara los resultados de la función `genre_weekday()`para Springfield y Shelbyville el lunes por la mañana (de 7 a 11) y el viernes por la tarde (de 17:00 a 23:00). Utilice el mismo formato de 24 horas que el conjunto de datos (por ejemplo, 05:00 = 17:00:00):

In [33]:
# llamando a la función para el lunes por la mañana en Springfield (utilizando spr_general en vez de la tabla df)
data= spr_general
day= 'Monday'
time1= '07:00:00'
time2= '11:00:00'

genre_weekday(data,day,time1,time2)

TypeError: '>' not supported between instances of 'SeriesGroupBy' and 'str'

In [ ]:
# llamando a la función para el lunes por la mañana en Shelbyville (utilizando shel_general en vez de la tabla df)
data= shel_general
day= 'Monday'
time1= '07:00:00'
time2= '11:00:00'

genre_weekday(data,day,time1,time2)

In [ ]:
# llamando a la función para el viernes por la tarde en Springfield
data= spr_general
day= 'Friday'
time1= '17:00:00'
time2= '23:00:00'

genre_weekday(data,day,time1,time2)

In [ ]:
# llamando a la función para el viernes por la tarde en Shelbyville
data= shel_general
day= 'Friday'
time1= '17:00:00'
time2= '23:00:00'
genre_weekday(data,day,time1,time2)

**Conclusión**

Por la noche solo 1 genero es diferente mientras que por la mañana solo hay 2 generos de diferencia entre ambas ciudades, rechazaría la hipotesis. 


<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~En general, realizas buenas conclusiones y buen código; sin embargo, hay cosas en la función ```genre_weekday``` que pueden mejorar. Primero que todo, los filtros que estás realizando para indexar no son consecutivos; podrían ser consecutivos. La idea sería algo como la siguiente:~~
    
```python
df = df[(condicion_1) & (condicion_2) & (condicion_3)]
```
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo Juan.
</div>

[Volver a Contenidos](#back)

### Hipótesis 3: preferencias de género en Springfield y Shelbyville <a id='genre'></a>

Hipótesis: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

Agrupa la tabla `spr_general` por género y encuentra el número de canciones reproducidas de cada género con el método `count()`. Después, ordena el resultado en orden descendente y guárdalo en `spr_genres`.

In [ ]:

#Corrección primera revisión
#Se asigna una columna especifica al groupby para evitar datos redundantes
spr_genres = spr_general.groupby('genre')['track'].count().sort_values(ascending=False)

# print(spr_genres)

Muestra las 10 primeras filas de `spr_genres`:

In [ ]:
# mostrar las 10 primeras filas de spr_genres
print(spr_genres.head(10))

Ahora haz lo mismo con los datos de Shelbyville.

Agrupa la tabla `shel_general` por género y encuentra el número de canciones reproducidas de cada género. Después, ordena el resultado en orden descendente y guárdalo en la tabla `shel_genres`:


In [ ]:

#Corrección primera revisión
#Se asigna una columna especifica al groupby para evitar datos redundantes
shel_genres = shel_general.groupby('genre')['track'].count().sort_values(ascending=False)


Muestra las 10 primeras filas de `shel_genres`:

In [ ]:
# Muestra las 10 primeras filas de shel_genres
print(shel_genres.head(10))

**Conclusión**

Hipótesis: Shelbyville ama la música rap. A los residentes de Springfield les gusta más el pop.

Falso, la música preferida en ambas ciudades es el pop, el genero rap no forma parte del top 10 de favoritos en ninguna de las ciudades. Las categoría rusrap entra dentro del top 10 en las últimas 3 posiciones para ambas ciudades.

[Volver a Contenidos](#back)

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
~~Fijate en como estas realizando los agrupamientos en ``spr_genres``, si seleccionas una varible usando el ``groupby`` podras evitar que se genere toda una tabla con informacion redudante. Te dare un ejemplo.~~
    
```python
df.groupby('city')['track'].count()   
```
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor #2</b> <a class="tocSkip"></a>
    
Buen trabajo Juan.
</div>

# Conclusiones <a id='end'></a>

`Conclusiones`


1. La actividad de los usuarios y las usuarias difiere según el día de la semana y dependiendo de la ciudad.
- __Aprobada__: Los días de mayor consumo de canciones para Shelbyville son los día miercoles, mientras que para springfield los días de mayor consumo suelen efectuarse los viernes, por lo tanto los días de consumo son diferentes en estas ciudades 
 
2. Los lunes por la mañana, los habitantes de Springfield y Shelbyville escuchan géneros distintos. Lo mismo ocurre los viernes por la noche.
- __Rechazada__: Por la noche solo 1 genero es diferente mientras que por la mañana solo hay 2 generos de diferencia entre ambas ciudades.
 
3. Los oyentes de Springfield y Shelbyville tienen preferencias distintas. En Springfield prefieren el pop, mientras que en Shelbyville hay más personas a las que les gusta el rap.
- __Rechazada__: La música preferida en ambas ciudades es el pop, el genero rap no forma parte del top 10 de favoritos en ninguna de las ciudades. Las categoría rusrap entra dentro del top 10 en las últimas 3 posiciones para ambas ciudades.


<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
Buen trabajo Juam.
</div>

### Nota
En proyectos de investigación reales, la prueba de hipótesis estadística es más precisa y cuantitativa. También ten en cuenta que no siempre se pueden sacar conclusiones sobre una ciudad entera a partir de datos de una sola fuente.

Aprenderás más sobre la prueba de hipótesis en el sprint de análisis estadístico de datos.

<div class="alert alert-block alert-danger">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General
    
~~Hola Juan.Te felicito por el desarrollo del proyecto hasta el momento, ahora bien, debes tener en cuenta los comentarios que he generado para la próxima entrega.~~
</div>

<div class="alert alert-block alert-success">
<b>Comentario del revisor</b> <a class="tocSkip"></a>
    
# Comentario General #2
    
Hola Juan.Te felicito por la culminación del proyecto. Muy buen trabajo.



[Volver a Contenidos](#back)